server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -389,6 +390,40 @@ public void onStoreClosed(ShardId shardId) {
         }
     }
 
+    public <T, E extends Exception> T withTempIndexService(final IndexMetadata indexMetadata,
+                                                           CheckedFunction<IndexService, T, E> indexServiceConsumer) throws IOException, E {
+        final Index index = indexMetadata.getIndex();
+        if (hasIndex(index)) {
+            throw new ResourceAlreadyExistsException(index);
+        }
+        List<IndexEventListener> finalListeners = Collections.singletonList(
+            // double check that shard is not created.
+            new IndexEventListener() {
+                @Override
+                public void beforeIndexShardCreated(ShardId shardId, Settings indexSettings) {
+                    assert false : "temp index should not trigger shard creation";
```

Comment:
minor: do we need those for easier test debugging?

Reply:
Not related to the tests. They make sure that the temporary index creation doesn't result in more work than expected. The shards mustn't get created as the temporary index service is only for change validation/processing.

Reply:
but the `assert` works only for the tests, that's why I'm asking.